populator-machinery/controller_test.go

```diff
@@ -0,0 +1,538 @@
+	testApiVersion         = testApiGroup + "/v1alpha1"
+	testDatasourceKind     = "TestKind"
+	testDataSourceName     = "test-data-source-name"
+	testStorageClassName   = "test-sc"
+	testPopulatorPvcName   = populatorPvcPrefix + "-" + testPvcUid
+	testPvName             = "test-pv"
+	testNodeName           = "test-node-name"
+	testPodName            = populatorPodPrefix + "-" + testPvcUid
+)
+
+var (
+	gvr = schema.GroupVersionResource{
+		Group:    testApiGroup,
+		Version:  "v1alpha1",
+		Resource: "testdatasources",
+	}
+	gk = schema.GroupKind{
+		Group: testApiGroup,
+		Kind:  testDatasourceKind,
+	}
+
+	kubeClient    = kubefake.NewSimpleClientset()
+	dynClient     = dynamicfake.NewSimpleDynamicClient(runtime.NewScheme())
+	gatewayClient = gatewayfake.NewSimpleClientset()
+
+	kubeInformerFactory = kubeinformers.NewSharedInformerFactory(kubeClient, time.Second*30)
+	dynInformerFactory  = dynamicinformer.NewDynamicSharedInformerFactory(dynClient, time.Second*30)
+
+	pvcInformer  = kubeInformerFactory.Core().V1().PersistentVolumeClaims()
+	pvInformer   = kubeInformerFactory.Core().V1().PersistentVolumes()
+	podInformer  = kubeInformerFactory.Core().V1().Pods()
+	scInformer   = kubeInformerFactory.Storage().V1().StorageClasses()
+	unstInformer = dynInformerFactory.ForResource(gvr).Informer()
+
+	gatewayInformerFactory = gatewayInformers.NewSharedInformerFactory(gatewayClient, time.Second*30)
+	referenceGrants        = gatewayInformerFactory.Gateway().V1beta1().ReferenceGrants()
+
+	populatorArgs = func(b bool, u *unstructured.Unstructured) ([]string, error) {
+		var args []string
+		return args, nil
+	}
+)
+
+func pvc(name, namespace, nodeName, scName, volumeName string, datasourceRef *v1.TypedObjectReference, phase v1.PersistentVolumeClaimPhase) *v1.PersistentVolumeClaim {
+	return &v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			UID:       testPvcUid,
+			Annotations: map[string]string{
+				annSelectedNode: nodeName,
+			},
+			Finalizers: []string{"kubernetes.io/pvc-protection"},
+		},
+		Spec: v1.PersistentVolumeClaimSpec{
+			AccessModes: []v1.PersistentVolumeAccessMode{
+				v1.ReadWriteOnce,
+			},
+			StorageClassName: &scName,
+			VolumeName:       volumeName,
+			DataSourceRef:    datasourceRef,
+		},
+		Status: v1.PersistentVolumeClaimStatus{
+			Phase: phase,
+		},
+	}
+}
+
+func dsf(apiGp, kind, name, namespace string) *v1.TypedObjectReference {
+	return &v1.TypedObjectReference{
+		APIGroup:  &apiGp,
+		Kind:      kind,
+		Name:      name,
+		Namespace: &namespace,
+	}
+}
+
+func ust() *unstructured.Unstructured {
+	return &unstructured.Unstructured{
+		Object: map[string]any{
+			"apiVersion": testApiVersion,
+			"kind":       testDatasourceKind,
+			"metadata": map[string]any{
+				"name":      testDataSourceName,
+				"namespace": testPvcNamespace,
+			},
+		},
+	}
+}
+
+func sc() *storagev1.StorageClass {
+	sc := testStorageClassName
+	p := "test.provisioner"
```

Comment:
let's use a variable for the provisioner name too?

Reply:
Done